### ISSUE TITLE
Performance Regression: mixer cpu use significantly higher in 1.0.3 

### ISSUE THREAD

**Issue by Stono** [istio/istio#9604]
Hey, 
Noticed this this morning.  In one of our production clusters on 1.0.2 (actually getting 10x the load of pre-prod):

Istio policy is nice and low
<img width="800" alt="screen shot 2018-10-29 at 13 26 35" src="https://user-images.githubusercontent.com/1486729/47652676-73cd5980-db7e-11e8-97b7-554feb0120c4.png">

And then our 1.0.3 upgraded pre-prod cluster, istio-policy using 15x as much:
<img width="801" alt="screen shot 2018-10-29 at 13 26 51" src="https://user-images.githubusercontent.com/1486729/47652706-8a73b080-db7e-11e8-8445-0bd1dfccd91c.png">

**Comment by Stono** [comment]
This is the same for telemetry (mixer based) too

**Comment by mandarjog** [comment]
Can you export your grafana mixer dashboard ( to raintank.io)?
Is there any other config change in your env?

**Comment by Stono** [comment]
Unfortunately not; happy to give your google account access to the preprod cluster though if you like (message me on slack).

Here are some screenshots:

<img width="1610" alt="screen shot 2018-10-29 at 15 50 51" src="https://user-images.githubusercontent.com/1486729/47662302-91a4b980-db92-11e8-95c2-ba68057bc1ce.png">
<img width="1602" alt="screen shot 2018-10-29 at 15 50 56" src="https://user-images.githubusercontent.com/1486729/47662303-91a4b980-db92-11e8-8f80-15b4b339f0bc.png">
<img width="1593" alt="screen shot 2018-10-29 at 15 51 00" src="https://user-images.githubusercontent.com/1486729/47662304-91a4b980-db92-11e8-97a9-6181bcda187f.png">
<img width="1613" alt="screen shot 2018-10-29 at 15 51 05" src="https://user-images.githubusercontent.com/1486729/47662305-923d5000-db92-11e8-9802-a5eefdb0698f.png">

**Comment by Stono** [comment]
Dunno if it helps but there are some errors in the logs:

**telemetry**
```
2018-10-29T15:48:49.789942Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:48:49.791311Z	info	Built new config.Snapshot: id='0'
2018-10-29T15:48:49.791575Z	info	Cleaning up handler table, with config ID:-1
2018-10-29T15:48:49.803809Z	warn	Failed to discover kinds: [statsd], start retry in background
2018-10-29T15:48:49.809188Z	info	Built new config.Snapshot: id='1'
2018-10-29T15:48:49.811538Z	info	Cleaning up handler table, with config ID:0
2018-10-29T15:48:49.811605Z	info	Awaiting for config store sync...
2018-10-29T15:48:50.814344Z	info	Publishing 8 events
2018-10-29T15:48:50.814699Z	error	failed to evaluate expression for field 'Variables[sourceWorkload]'; unknown attribute source.workload.name
2018-10-29T15:48:50.814768Z	error	failed to evaluate expression for field 'Variables[sourcePrincipal]'; unknown attribute source.principal
2018-10-29T15:48:50.814799Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:50.814810Z	error	No valid actions found in rule
2018-10-29T15:48:50.814829Z	error	unknown attribute context.protocol
2018-10-29T15:48:50.814850Z	error	Instance not found: instance='requestcount.metric'
2018-10-29T15:48:50.814867Z	error	Instance not found: instance='requestduration.metric'
2018-10-29T15:48:50.814876Z	error	Instance not found: instance='requestsize.metric'
2018-10-29T15:48:50.814881Z	error	Instance not found: instance='responsesize.metric'
2018-10-29T15:48:50.814935Z	error	No valid instances found
2018-10-29T15:48:50.814947Z	error	No valid actions found in rule
2018-10-29T15:48:50.814960Z	error	unknown attribute context.protocol
2018-10-29T15:48:50.814969Z	error	Instance not found: instance='tcpbytesent.metric'
2018-10-29T15:48:50.814977Z	error	Instance not found: instance='tcpbytereceived.metric'
2018-10-29T15:48:50.814981Z	error	No valid instances found
2018-10-29T15:48:50.814988Z	error	No valid actions found in rule
2018-10-29T15:48:50.815001Z	error	unknown attribute context.protocol
2018-10-29T15:48:50.815012Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:50.815021Z	error	No valid actions found in rule
2018-10-29T15:48:50.815026Z	info	Built new config.Snapshot: id='2'
2018-10-29T15:48:50.815100Z	info	Cleaning up handler table, with config ID:1
2018-10-29T15:48:52.009585Z	info	Publishing 1 events
2018-10-29T15:48:52.011153Z	error	error evaluating AttributeBinding expression for attribute key 'destination.namespace': unknown attribute destination.namespace
2018-10-29T15:48:52.011218Z	error	failed to evaluate expression for field 'Variables[requestSize]'; unknown attribute request.size
2018-10-29T15:48:52.011306Z	error	failed to evaluate expression for field 'Variables[connection_security_policy]'; unknown attribute context.reporter.kind
2018-10-29T15:48:52.011327Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:52.011342Z	error	No valid actions found in rule
2018-10-29T15:48:52.011367Z	error	unknown attribute context.protocol
2018-10-29T15:48:52.011381Z	error	Instance not found: instance='requestcount.metric'
2018-10-29T15:48:52.011386Z	error	Instance not found: instance='requestduration.metric'
2018-10-29T15:48:52.011391Z	error	Instance not found: instance='requestsize.metric'
2018-10-29T15:48:52.011402Z	error	Instance not found: instance='responsesize.metric'
2018-10-29T15:48:52.011407Z	error	No valid instances found
2018-10-29T15:48:52.011415Z	error	No valid actions found in rule
2018-10-29T15:48:52.011427Z	error	unknown attribute context.protocol
2018-10-29T15:48:52.011443Z	error	Instance not found: instance='tcpbytesent.metric'
2018-10-29T15:48:52.011452Z	error	Instance not found: instance='tcpbytereceived.metric'
2018-10-29T15:48:52.011457Z	error	No valid instances found
2018-10-29T15:48:52.011461Z	error	No valid actions found in rule
2018-10-29T15:48:52.011478Z	error	unknown attribute context.protocol
2018-10-29T15:48:52.011528Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:52.011533Z	error	No valid actions found in rule
2018-10-29T15:48:52.011537Z	info	Built new config.Snapshot: id='3'
2018-10-29T15:48:52.011606Z	info	Cleaning up handler table, with config ID:2
2018-10-29T15:48:53.011837Z	info	Publishing 8 events
2018-10-29T15:48:53.013381Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:53.013420Z	error	No valid actions found in rule
2018-10-29T15:48:53.013429Z	error	Handler not found: handler='handler.kubernetesenv'
2018-10-29T15:48:53.013432Z	error	No valid actions found in rule
2018-10-29T15:48:53.013461Z	info	Built new config.Snapshot: id='4'
2018-10-29T15:48:53.014186Z	info	adapters	serving prometheus metrics on 42422	{"adapter": "handler.prometheus.istio-system"}
2018-10-29T15:48:53.015295Z	info	Cleaning up handler table, with config ID:3
2018-10-29T15:48:54.206963Z	info	Publishing 1 events
2018-10-29T15:48:54.210263Z	info	Built new config.Snapshot: id='5'
2018-10-29T15:48:54.210401Z	info	adapters	getting kubeconfig from: ""	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:54.210435Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:48:54.211404Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:54.611863Z	info	adapters	Cache sync successful.	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:54.614576Z	info	Cleaning up handler table, with config ID:4
2018-10-29T15:48:54.614622Z	error	adapters	adapter did not close all the scheduled daemons	{"adapter": "handler.prometheus.istio-system"}
2018-10-29T15:48:55.631855Z	info	Starting monitor server...
Istio Mixer: root@0ead81bba27d-docker.io/istio-1.0.3-a44d4c8bcb427db16ca4a439adfbd8d9361b8ed3-Clean
Starting gRPC server on port 9091
2018-10-29T15:48:55.690412Z	info	ControlZ available at 10.194.6.20:9876
2018-10-29T15:49:49.124672Z	info	Publishing 18 events
2018-10-29T15:49:49.127003Z	info	Built new config.Snapshot: id='6'
2018-10-29T15:49:49.127106Z	info	adapters	getting kubeconfig from: ""	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.127119Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:49:49.128236Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.533702Z	info	adapters	Cache sync successful.	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.537608Z	info	Cleaning up handler table, with config ID:5
2018-10-29T15:49:49.537675Z	error	adapters	adapter did not close all the scheduled daemons	{"adapter": "handler.kubernetesenv.istio-system"}
```

**policy**
```

2018-10-29T15:48:40.956715Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:48:40.957780Z	info	Built new config.Snapshot: id='0'
2018-10-29T15:48:40.957930Z	info	Cleaning up handler table, with config ID:-1
2018-10-29T15:48:41.015211Z	warn	Failed to discover kinds: [statsd], start retry in background
2018-10-29T15:48:41.015555Z	info	Built new config.Snapshot: id='1'
2018-10-29T15:48:41.015726Z	info	Cleaning up handler table, with config ID:0
2018-10-29T15:48:41.015738Z	info	Awaiting for config store sync...
2018-10-29T15:48:42.047822Z	info	Publishing 9 events
2018-10-29T15:48:42.048927Z	info	Built new config.Snapshot: id='2'
2018-10-29T15:48:42.049001Z	info	Cleaning up handler table, with config ID:1
2018-10-29T15:48:43.193015Z	info	Publishing 3 events
2018-10-29T15:48:43.194772Z	info	Built new config.Snapshot: id='3'
2018-10-29T15:48:43.194859Z	info	Cleaning up handler table, with config ID:2
2018-10-29T15:48:44.393184Z	info	Publishing 1 events
2018-10-29T15:48:44.395259Z	info	Built new config.Snapshot: id='4'
2018-10-29T15:48:44.395343Z	info	Cleaning up handler table, with config ID:3
2018-10-29T15:48:46.193390Z	info	Publishing 5 events
2018-10-29T15:48:46.195448Z	info	Built new config.Snapshot: id='5'
2018-10-29T15:48:46.195588Z	info	adapters	getting kubeconfig from: ""	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:46.195604Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:48:46.196816Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:46.924202Z	info	Starting monitor server...
Istio Mixer: root@0ead81bba27d-docker.io/istio-1.0.3-a44d4c8bcb427db16ca4a439adfbd8d9361b8ed3-Clean
Starting gRPC server on port 9091
2018-10-29T15:48:46.961568Z	info	ControlZ available at 10.194.24.28:9876
2018-10-29T15:48:46.997057Z	info	adapters	Cache sync successful.	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:48:46.997866Z	info	adapters	serving prometheus metrics on 42422	{"adapter": "handler.prometheus.istio-system"}
2018-10-29T15:48:47.000454Z	info	Cleaning up handler table, with config ID:4
2018-10-29T15:49:49.122914Z	info	Publishing 18 events
2018-10-29T15:49:49.124719Z	info	Built new config.Snapshot: id='6'
2018-10-29T15:49:49.124942Z	info	adapters	getting kubeconfig from: ""	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.125054Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-10-29T15:49:49.125958Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.426252Z	info	adapters	Cache sync successful.	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.428849Z	info	Cleaning up handler table, with config ID:5
2018-10-29T15:49:49.428917Z	error	adapters	adapter did not close all the scheduled daemons	{"adapter": "handler.kubernetesenv.istio-system"}
2018-10-29T15:49:49.428931Z	error	adapters	adapter did not close all the scheduled daemons	{"adapter": "handler.prometheus.istio-system"}
```

**Comment by mandarjog** [comment]
Incoming requests into telemetry and dispatch count have become 8x.
And cpu has also gone up, so telemetry is doing more work.
What is the total mesh traffic? Has size of deployment changed at all ?

**Comment by Stono** [comment]
No, as I keep saying mate nothing at all had changed other than an upgrade of control plane and sidecars to 1.0.3

**Comment by Stono** [comment]
Could this be istio-proxy 1.0.3 just making a lot more policy calls than 1.0 2, so the issue could be with proxy?

**Comment by mandarjog** [comment]
There is a change in client side caching code in 1.0.3.
@JimmyCYJ @qiwzhang We are observing higher inbound requests into mixer with absolutely no other change in deployment beyond upgrading the control plane. Can you check if change in batch size + change in cache code results in 8x traffic to both istio-proxy and istio-telemetry.

**Comment by qiwzhang** [comment]
Just check the proxy commit SHA.  There is little or not change between proxy 1.02 and 1.03.  This is the only change:  https://github.com/istio/proxy/pull/1969

**Comment by jnativio** [comment]
Note that regpatrol detected this problem with the master branch and not in the 1.0 branch where 1.0.3 was cut from. The  Pilot CPU and memory usage were normal for recent builds of 1.0 branch. See for example: 
https://ibmcloud-perf.istio.io/regpatrol/release-1.0-20181029-09-15/istio_full/

**Comment by mandarjog** [comment]
@Stono did you see CPU utilization go up as soon as the control plane was updated? That is Mixer at 1.0.3 but proxies still at the previous version?

**Comment by mandarjog** [comment]
Another related question: when you upgraded to 1.0.3 did you also do `kubectl apply -f crds.yaml` ?
I am thinking of a change between 0.2 and 0.3 that looks for uninitialized CRDs continuously. This is independent of the data path. If absence of the `kubectl apply`, we are missing two CRDs that are compiled into Mixer, specifically `kind: cloudwatch`, `kind: dogstatsd`. Mixer may continually look for new CRDs.

**Comment by Stono** [comment]
@mandarjog I am not sure if it was before or after proxy updates, I might have time to test again later today but I'm not sure.

We don't use kubectl apply -f to do upgrades, we use helm (if that makes any difference)

**Comment by stefanprodan** [comment]
I can confirm that the CPU usage goes back to normal after applying the 1.0.3 CRDs manually. Maybe there is a problem with Helm upgrade from 1.0.2 to 1.0.3

**Comment by mandarjog** [comment]
Thanks for confirming the work around. I will get a proper fix out soon.

@stono helm upgrade is not processing the crds correctly.

**Comment by jnativio** [comment]
I could not reproduce the problem with that master build or more recent ones. Regpatrol did not generate the flame graph for istio-policy during the run, but there is the log saved for this component (https://ibmcloud-perf.istio.io/regpatrol/master-20181101-09-15/istio_full/istio-policy-68d97879dc-bdxh4.log), where this error msg appears continuously:

_2018-11-02T08:47:47.625333Z	error	attributes	string attribute context.reporter.kind: string index 219 is not defined in the available dictionaries_

There is also the profile dump file --> https://ibmcloud-perf.istio.io/regpatrol/master-20181101-09-15/istio_full/aaa_policy_prof.dump

**Comment by Stono** [comment]
@jnativio I didn't see that in my policy logs! 

@stefanprodan it'd be great if you could diff your applied resources before and after to know what has changed?  When I do a `helm template` on 1.0.2 vs 1.0.3 and diff I'm not seeing *any* CRD changes?  @mandarjog what do you believe has changed? 

@sdake are you aware of any helm issues going from 1.0.2 -> 1.0.3?  I know you've been keen to ensure upgradability.

**Comment by Stono** [comment]
These are the only two changes I see between 1.0.2 and 1.0.3 when it comes to CRD definitions:

```
diff --git a/istio/templates/vendor/custom-resources/cloudwatches.config.istio.io.yaml b/istio/templates/vendor/custom-resources/cloudwatches.config.istio.io.yaml
new file mode 100644
index 0000000..b8ab2c1
--- /dev/null
+++ b/istio/templates/vendor/custom-resources/cloudwatches.config.istio.io.yaml
@@ -0,0 +1,19 @@
+---
+kind: CustomResourceDefinition
+apiVersion: apiextensions.k8s.io/v1beta1
+metadata:
+  name: cloudwatches.config.istio.io
+  annotations: {}
+  labels:
+{{ include "istio.labels.standard" . | indent 4 }}
+spec:
+  group: config.istio.io
+  names:
+    kind: cloudwatch
+    plural: cloudwatches
+    singular: cloudwatch
+    categories:
+    - istio-io
+    - policy-istio-io
+  scope: Namespaced
+  version: v1alpha2
diff --git a/istio/templates/vendor/custom-resources/dogstatsds.config.istio.io.yaml b/istio/templates/vendor/custom-resources/dogstatsds.config.istio.io.yaml
new file mode 100644
index 0000000..a8cd3c1
--- /dev/null
+++ b/istio/templates/vendor/custom-resources/dogstatsds.config.istio.io.yaml
@@ -0,0 +1,19 @@
+---
+kind: CustomResourceDefinition
+apiVersion: apiextensions.k8s.io/v1beta1
+metadata:
+  name: dogstatsds.config.istio.io
+  annotations: {}
+  labels:
+{{ include "istio.labels.standard" . | indent 4 }}
+spec:
+  group: config.istio.io
+  names:
+    kind: dogstatsd
+    plural: dogstatsds
+    singular: dogstatsd
+    categories:
+    - istio-io
+    - policy-istio-io
+  scope: Namespaced
+  version: v1alpha2
```

I do not see *any* changes to CRD implementations?   

What exactly do people believe that helm is not applying correctly?

**Comment by stefanprodan** [comment]
@Stono this two CRDs were missing after I've upgraded to 1.0.3

```
customresourcedefinition.apiextensions.k8s.io/cloudwatches.config.istio.io created
customresourcedefinition.apiextensions.k8s.io/dogstatsds.config.istio.io created
```

I've used kubectl to apply the CRDs and killed all Mixer pods, after the "restart" the CPU usage went back to normal.

**Comment by Stono** [comment]
So I *did* have those CRDs (I don't do a straight up helm install) and still saw high CPU... so something else is wrong.

**Comment by mandarjog** [comment]
@Stono helm is not doing equivalent of `kubectl apply -f crds.yaml`. 
In your logs, `statsd` is the kind that is missing, which is *not new* in 1.0.3

**Comment by mandarjog** [comment]
@Stono if you can test again today, it would be great to apply crds manually.

**Comment by Stono** [comment]
I'm sorry but I'm really not following this at all.

In Istio 1.0.2, on our cluster, we do not have those CRDs:

```
❯ k get crd | grep -i cloud
backendconfigs.cloud.google.com               2018-11-05T12:48:21Z
servicechecks.atcloud.io                      2018-10-19T14:22:04Z
services.atcloud.io                           2018-07-08T11:33:53Z
```

Istio 1.0.3 it appears to add `cloudwatches.config.istio.io`, are you saying it shouldn't?

**Comment by Stono** [comment]
Ahh you're saying that my missing of the `statsd` CRD is causing the problem?  I removed the statsd stuff manually because of https://github.com/istio/istio/issues/7874

**Comment by mandarjog** [comment]
Any missing CRDs can cause high CPU problem in Mixer 1.0.3.
1.0.3 added  2 new CRDs `kind: cloudwatch, kind: dogstatsd`. 
You need to add `statsd` CRD and we need to figure out why helm did not keep it in sync.

**Comment by Stono** [comment]
OK that makes more sense, thanks.

**Comment by mandarjog** [comment]
Ok great. I will have a fix out soonish.

**Comment by williamchong007** [comment]
FYI I was investigating a perf issue in `1.0.2` where mixer takes up 2000% CPU utilization, reapplying `crds.yml` solved my issue.

I think I installed `1.0` by helm then upgraded to `1.0.2`, so this issue might be from a previous version.

**Comment by sdake** [comment]
@stefanprodan @williamchong007 unfortunately Helm does not properly handle the upgrade case when new CRDs are introduced into the system.  This is tied up in a problem with  the Helm job `crd-install` not actually installing new CRDs during an upgrade process.  Helm 2.10+ treats CRDs as "unmanaged objects" during the install process - in other words Helm doesn't track these CRDs via tiller.

The upgrade documentation needs an upgrade.  I'm not sure what @mandarjog had in mind for a solution.

I can't think of any other viable technique to handle CRD creation other than creating a process/container that installs and or upgrades the CRDs as a job.

My initial solution for 1.0.4 is as follows:
1. break out the CRDs into a separate chart which must be installed independently.
1. Remove `crd-install` as a feature - until either Helm solves this problem upstream or we find a more complete solution possibly via an operator.
1. Enforce a two phase Helm installation process.
  1. Install/upgrade CRDs helm chart.
  1. Install/upgrade the Istio helm chart.

We have used post-install jobs for this in the past, and found that model to be untenable.

I'll need to test the upgrade scenarios with this new model.  Will submit a PR this morning.

Cheers
-steve

**Comment by sdake** [comment]
See https://github.com/helm/helm/pull/3982 for more details.

**Comment by sdake** [comment]
https://github.com/istio/istio/pull/9847

**Comment by munnerz** [comment]
Subscribing to this - we have a similar issue with cert-manager too.

To roughly summarise the current position, which I think applies to Istio too:

* Users upgrading from a release *without* the annotation see all their CRDs deleted (which is really, really bad)
* Users who upgrade and then manually fix it (by manually deleting/creating CRDs), will not be able to upgrade if we *remove* the annotation
* Users who install for the first time *with* the annotation will not be able to upgrade if we *remove* it, and won’t get updates to CRDs if we *keep* it

I'm hoping to hear back from the wider Helm community on what a recommended path to fixing this is - I think a 2-phase Helm install may be required.

**Comment by Stono** [comment]
At AutoTrader we opt for two phased deployments (managed with helmfile)
because of this (not just istio)

**Comment by sdake** [comment]
@munnerz your further along than we are in understanding this problem.  For your third bullet is there any workaround?

My current recommended approach for proceeding is to immediately remove the `crd-install` annotation since it completely blocks upgrades.

This will leave some fallout for operators that have installed with the `crd-install` annotation in the pastl.  I'm really not sure what to tell our operators that have used `crd-install` with Helm `2.10.0` already - as it appears they have no upgrade path.

@costinm ^^

Cheers
-steve

**Comment by Stono** [comment]
I can't see any upgrade path either at the moment either, it's a bit of a mess.  Even if you separate it out into two charts (eg install CRD definitions first) then unless the previously manually added definitions (eg kubectl apply -f, or helms crd-install hook) are manually remove first, it'll fail.

However, manually removing those CRD's for people will also remove all of their implements (VirtualServices, DestinationRules etc).  Which is a no go either as @munnerz said - really really bad.

I see nothing but disruptive upgrade paths. 

I'm wondering if it's worth tracking this in another issue now perhaps as this one was about the mixer load which was identified as it constantly scanning for CRDs, feels like a separate issue that upgrades are broken?

**Comment by sdake** [comment]
@mandarjog - I offered to resolve this PR thinking it was solvable by breaking out the charts.  I suspect you are adding a fail-open retry to the mixer crd detection loop.  I'll leave you to it as I've got all day commitments today->Wednesday.

Cheers
-steve

**Comment by mandarjog** [comment]
I am going to fix the issue by 
1. Mixer will wait more efficiently on the crds
2. It will not wait for ever

**Comment by costinm** [comment]
I think I have a way to fix this - I was planning to push for it in 1.2, but maybe we can do parts of it sooner.

Last env meetings we had pretty good agreement to introduce an 'operator' and reorganize the install/upgrade and configs around features ( telemetry, control plane, etc ) instead of components ( mixer, pilot ).  We also have strong security reasons to better isolate the istio components - an admin making 
changes to our grafana templates should not have access to Citadel root private key for example. I did a bit of prototyping, and it appears possible to install separate namespaces for pilot, ingress, egress, telemetry - and even run different versions at the same time, so people upgrading could test the new 
version along the old one ( we call it 'canary' ), without disruption. It already works with the code we have, 
using annotations on the pod to select a different pilot namespace for some pods - but with small changes 
to sidecar injector we can make it much simpler. 

So my proposal will be to keep only Citadel ( which is very hard to move out ) and possibly Galley in istio-system, with no custom configs using our CRDs. In addition we are already planning to have an operator - 
the first job could be to maintain the CRDs. That means no change to the existing 1.0 helm or crd file - so nothing will be broken ( and nothing will be added there ). Pilot already has code to auto-create CRDs, it can be moved to a separate component, or Galley could do it if we keep it in istio-system. 

I'll send an email to istio-dev or discuss this in next env meeting - my original motivation was improving security and allowing users to test a new release in parallel with the old one.

**Comment by mandarjog** [comment]
Addressed by #9922
